Documentation revamp (#623)
dapphub/dapptools: src/dapp/README.md

property based testing, but is also more difficult, especially for complicated functions.

Continuing with our vault example, imagine that we forgot the password and did not have the source available. 
We can symbolically explore all possibilities to find the one that lets us withdraw by writing a `proveFail` test:

```solidity
function proveFail_withdraw(uint guess) public {
    address(dapptutorial).transfer(1 ether);
    uint preBalance = address(this).balance;
    dapptutorial.withdraw(guess);
    uint postBalance = address(this).balance;
    assertEq(preBalance + 1 ether, postBalance);
}
```

When we run this with `dapp test`, we are given a counterexample:
```
Failure: proveFail_withdraw(uint256)

  Counterexample:

    result:   Successful execution
    calldata: proveFail_withdraw(42)
```

which demonstrates that if we give the password `42`, it is possible to withdraw from the vault.

For more reading on property based testing and symbolic execution, see [this tutorial on the Ethereum Foundation blog](https://fv.ethereum.org/2020/12/11/symbolic-execution-with-ds-test/).

> ### Testing against RPC state

Author: maybe not in this tutorial, but maybe in the expanded `dapp test` section

### `dapp test`

    Usage: dapp test [<options>]
  
    Options:
        -v, --verbose             trace ouput for failing tests
        -vv                       trace output for all tests including passes
        --verbosity <number>      sets the verbosity to <number>
        --fuzz-runs <number>      number of times to run fuzzing tests
        --replay <string>         rerun a particular test case
        -m, --match <string>      only run test methods matching regex

    RPC options:
        --rpc                     fetch remote state via ETH_RPC_URL
        --rpc-url <url>           fetch remote state via <url>
        --rpc-block <number>      block number (latest if not specified)

    SMT options:
        --smttimeout <number>     timeout passed to the smt solver in ms (default 30000)
        --solver <string>         name of the smt solver to use (either "z3" or "cvc4")
        --max-iterations <number> number of times we may revisit a particular branching point during symbolic execution

dapp tests are written in Solidity using the `ds-test` module. To install it, run
```sh
dapp install ds-test
```

> Every contract which inherits from `DSTest` will be treated as a test contract, and will be assumed to have a public `setUp()` function, which will be run before every test.

Author: correct

> Every contract which inherits from `DSTest` will be treated as a test contract, and will be assumed to have a public `setUp()` function, which will be run before every test.

Author: ```suggestion
Every contract which inherits from `DSTest` will be treated as a test contract, if it has a `setUp()` function, it will be run before every test.
```
